REGRESSION(286663@main): `export-w3c-test-changes` fails due to Bugzilla redesign

briannafan opened the pull request:
#### e49d922a53bd718e60e93fd8e5dccce9e1f134ab
<pre>
REGRESSION(286663@main): `export-w3c-test-changes` fails due to Bugzilla redesign
<a href="https://rdar.apple.com/143030227">rdar://143030227</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=293015">https://bugs.webkit.org/show_bug.cgi?id=293015</a>

Reviewed by Sam Sneddon.

Convert all bugzilla interactions to use webkitbugspy (Bugzilla REST API).

* Tools/Scripts/webkitpy/w3c/test_exporter.py:
(WebPlatformTestExporter.__init__):
   - We instantiate the Git repo in order to populate our issue trackers. It&apos;ll
     be used in future PRs to perform all git operations.
(WebPlatformTestExporter.make_pull_request):
(parse_args):
* Tools/Scripts/webkitpy/w3c/test_exporter_unittest.py:
(TestExporterTest):
(TestExporterTest.setUp):
(TestExporterTest.test_export):
(TestExporterTest.test_export_with_specific_branch):
(TestExporterTest.test_export_no_clean):
(TestExporterTest.test_export_interactive_mode):
(TestExporterTest.test_export_invalid_token):
(TestExporterTest.test_export_wrong_token):
(TestExporterTest.test_has_wpt_changes):
(TestExporterTest.test_has_no_wpt_changes_for_no_diff):

Canonical link: <a href="https://commits.webkit.org/295181@main">https://commits.webkit.org/295181@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8481989385ecd6c0ab2d00e64f35bf8bd948513

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104325 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24029 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14443 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/109526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/54993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106365 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24401 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32576 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/109526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/54993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107331 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/18961 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/94145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/109526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/103803 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/12217 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/54353 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/12273 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111907 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31482 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88250 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31846 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/90389 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/87914 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/32812 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10577 "Build is in progress. Recent messages:") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16938 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31410 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/31204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/34540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/32764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -51,34 +53,45 @@
+# TODO: Need to check credential management for webkitbugspy
 class WebPlatformTestExporter(object):
-    def __init__(self, host, options, gitClass=Git, bugzillaClass=Bugzilla, WPTGitHubClass=WPTGitHub, WPTLinterClass=WPTLinter):
+    def __init__(self, host, options, gitClass=Git, bugzillaClass=Bugzilla, WPTGitHubClass=WPTGitHub, WPTLinterClass=WPTLinter, repository=local.Git(os.path.abspath(os.getcwd()))):
```
We should default to `None` and then create the repo object within the constructor, because we do want separate instances each time, I'd presume?

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -51,34 +53,45 @@
-        self._gitClass = gitClass
-        self._bugzilla = bugzillaClass()
+        self._gitClass = gitClass  # FIXME: Move git operations to webkitscmpy
+        self._bugzilla = bugzillaClass
```
We're no longer instantiating the class? That seems like it will break our usage of `self._bugzilla` below?

briannafan replied:
Ah typo!

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -51,34 +53,45 @@
                 self._bug_id = self._host.checkout().bug_id_for_this_commit(options.git_commit)
 
+        if string_utils.decode(self._bug_id).isnumeric() and Tracker.instance():
+            issue = Tracker.instance().issue(int(self._bug_id))
```
We don't actually register a tracker anywhere, do we? `webkitscmpy.local.scm.Scm.__init__` does, but we haven't instantiated that or any subclass anywhere.

briannafan replied:
That's actually why I'm instantiating the repository even though this change doesn't use it. The trackers are populated with `local.Git` as well

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -52,33 +54,44 @@
         self._WPTGitHubClass = WPTGitHubClass
-        self._gitClass = gitClass
+        self._gitClass = gitClass  # FIXME: Move git operations to webkitscmpy
+        self.repository = repository or local.Git(os.path.abspath(os.getcwd()))
```
We use `local.Scm.from_path(self.host.filesystem.getcwd())` elsewhere, interestingly. I don't know how much we care about the filesystem abstraction here…

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -52,33 +54,44 @@
         if not self._bug_id:
             if options.attachment_id:
-                self._bug_id = self._bugzilla.bug_id_for_attachment_id(options.attachment_id)
+                self._bug_id = self._bugzilla.bug_id_for_attachment_id(options.attachment_id)  # FIXME: Dependency on webkitpy.bugzilla should be removed when patch workflow is disabled
```
We may even just want to disable the ability to export changes from patches, even before the workflow is totally gone, really.

briannafan replied:
Yeah, let's do this in a follow-up PR

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -323,8 +336,10 @@ def make_pull_request(self):
-            self._bugzilla.post_comment_to_bug(self._bug_id, "Submitted web-platform-tests pull request: " + pr_url, see_also=[pr_url])
+            pr_url = f'{WPT_PR_URL}{pr_number}'
+            # TODO: Waiting to land PR #45278 to call this function
+            # self._bug.add_related_links([pr_url])
```
This is a blocker, given https://github.com/web-platform-tests/wpt-pr-bot/blob/f03de165700a8dcb8ab47b0a15c8c0be9f083fe2/lib/metadata/webkit.js#L11-L13 verifies the link back here.

briannafan replied:
Landed! I added to the unit tests to check for this as well

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter_unittest.py:
```diff
@@ -20,18 +20,30 @@
+    def setUp(self):
+        super().setUp()
+        os.mkdir(os.path.join(self.path, '.git'))
+        os.mkdir(os.path.join(self.path, '.svn'))
```
I'd presume we don't need `.svn`?

gsnedders reviewed (changes requested):
A few nits; mostly r- because I want to make sure we land #45278 first.

briannafan committed:
REGRESSION(286663@main): `export-w3c-test-changes` fails due to Bugzilla redesign
rdar://143030227
https://bugs.webkit.org/show_bug.cgi?id=293015

Reviewed by Sam Sneddon.

Convert all bugzilla interactions to use webkitbugspy (Bugzilla REST API).

* Tools/Scripts/webkitpy/w3c/test_exporter.py:
(WebPlatformTestExporter.__init__):
   - We instantiate the Git repo in order to populate our issue trackers. It'll
     be used in future PRs to perform all git operations.
(WebPlatformTestExporter.make_pull_request):
(parse_args):
* Tools/Scripts/webkitpy/w3c/test_exporter_unittest.py:
(TestExporterTest):
(TestExporterTest.setUp):
(TestExporterTest.test_export):
(TestExporterTest.test_export_with_specific_branch):
(TestExporterTest.test_export_no_clean):
(TestExporterTest.test_export_interactive_mode):
(TestExporterTest.test_export_invalid_token):
(TestExporterTest.test_export_wrong_token):
(TestExporterTest.test_has_wpt_changes):
(TestExporterTest.test_has_no_wpt_changes_for_no_diff):

Canonical link: https://commits.webkit.org/295181@main
